modules/core/src/lib/layer.js

```diff
@@ -390,11 +390,24 @@ export default class Layer extends Component {
       ignoreUnknownAttributes: true
     });
 
-    const model = this.getSingleModel();
+    let model;
+
+    model = this.getSingleModel();
     if (model) {
       const changedAttributes = attributeManager.getChangedAttributes({clearChangedFlags: true});
       model.setAttributes(changedAttributes);
     }
+
+    if (props.modelAttributeUpdaters) {
```

Comment:
This logic is internal to the layer and should not be part of the props system. Props reflect user-customizable layer settings. Do you consider adding an extra descriptor to each layer attribute?

Reply:
I don't think there's a way to do everything that needs to be done with extra attribute properties. There are 4 things that happened in the custom `updateAttributes`:
- Filter certain attributes to one model or the other
- Switch attributes between instanced and non-instanced depending on the model.
- Update the vertexCount in one model
- Update the instanceCount in the other model

I don't love the idea of using callbacks to solve this problem, but I'm not sure how to do all of that without them. Would moving the callbacks out of `props` (maybe just making them member properties of the layer?) be enough?

Reply:
There are two different things we want to do here:
1. Upload new attribute values to the GPU. These can be expensive so we only want to do it when they are recalculated.
2. Set `vertexCount`/`instanceCount`. They are not tied to any particular attribute. They also cost virtually nothing.

I think it's fair to limit this exploration to 1.

Reply:
Actually, I have a new idea for this. I think the shaderAttributes + conditional compilation should be enough to handle at least filtering and switching b/w instanced and non-instanced. I don't think I can filter out the indices for the sideModel, though. I'll take a crack at this when I get into the office.